### PR TITLE
Merge live and model links

### PR DESCRIPTION
**Pull request by subsymbolic**
Reviewer: Caine and Chris
Asana: https://app.asana.com/0/414235014887631/469637695297538

**Description**:
Stop TabViewController from updating the model with blank fields in cases where the model has better data.

It is possible for the model to have a title from a previously persisted session and for the TabViewController's title to be blank while it is loading - so even though the TabViewController is most up-to-date we still want to fallback on the model data at times.

**Steps to test this PR**:
1. Open the app, launch a few tabs
2. Ensure they are fully loaded (check that all details are populated in the tab switcher)
3. Kill and open the app again
4. Open the tab switcher and ensure that all title data etc is populated.
5. Kill some tabs and ensure that the tab switcher data is maintained as items are reloaded from the cache.

###### Reviewer Checklist:
- [ ] **Ensure the PR solves the problem**
- [ ] **Review every line of code**
- [ ] **Ensure the PR does no harm by testing the changes thoroughly**
- [ ] **Get help if you're uncomfortable with any of the above!**
- [ ] Determine if there are any quick wins that improve the implementation

###### PR DRI Checklist:
- [x] Get advice or leverage existing code
- [x] Agree on technical approach with reviewer (if the changes are nuanced)
- [x] Ensure that there is a testing strategy (and documented non-automated tests)
- [x] Ensure there is a documented monitoring strategy (if necessary)
- [x] Consider systems implications (Database connections, Grafana stats, CPU)

### CODE REVIEW

File: DuckDuckGo/TabViewController.swift [hunk 1]
```diff
@@ -48,6 +48,15 @@ class TabViewController: WebViewController {
         webEventsDelegate = self
     }
     
+    public var link: Link? {
+        guard let url = url else {
+            return Link.mergeData(primary: nil, secondary: tabModel.link)
```
Comment (brindy) [inline DuckDuckGo/TabViewController.swift]: Given this function just returns the secondary if the primary is nil anyway, could just return the `tabModel.link` in this case and simplify `mergeData`?

File: DuckDuckGo/TabViewController.swift [hunk 2]
```diff
@@ -48,6 +48,15 @@ class TabViewController: WebViewController {
...
+        }
+        
+        let activeLink = Link(title: name, url: url, favicon: favicon)
+        return Link.mergeData(primary: activeLink, secondary: tabModel.link)
+    }
```
Comment (brindy) [inline DuckDuckGo/TabViewController.swift]: Given primary will never be nil in this case, `mergeData` could just be an instance method which returns a new Link when merged, e.g.

`return activeLink.merge(with: tabModel.link)`

File: Core/Link.swift [hunk 3]
```diff
@@ -59,4 +59,26 @@ public class Link: NSObject, NSCoding {
...
+     */
+    public static func mergeData(primary: Link?, secondary: Link?) -> Link? {
+        guard let primary = primary else {
+            return secondary
+        }
+        
+        guard let secondary = secondary else {
+            return primary
+        }
+
```
Comment (brindy) [inline Core/Link.swift]: This should probably be an instance method (see comments below) in which case these guards would also be unnecessary.

**Comment by subsymbolic** [comment]
@brindy, I've switched the static over to an instance method. Ready for rereview!

**Comment by brindy** [comment]
LGTM, thanks @subsymbolic

**Pull request merged**